### ISSUE TITLE
[caffe2] VS2017 compiler error for ATen

### ISSUE THREAD

**Issue by ygwangmsra** [pytorch/pytorch#12161]
VS2017, (v141, update 6), compiler 1913, error for 
"ATen/Native/cpu" cpp file: GridSamplerKernel.cpp.AVX2.cpp 

However, when I use the compiler1900 (VS2015, v140)， there is no error, which can be sucessfully compiled.

Hope to fix the bug.

**Comment by peterjc123** [comment]
@ygwangmsra Could you post the full log here?

**Comment by ygwangmsra** [comment]
That is it.  @peterjc123

![error](https://user-images.githubusercontent.com/43432586/46200735-ffb45180-c344-11e8-8e31-67da52bc472e.png)

Thanks!

**Comment by ygwangmsra** [comment]
And h
[build.log](https://github.com/pytorch/pytorch/files/2427781/build.log)
ere is the full log. @peterjc123

**Comment by peterjc123** [comment]
@ygwangmsra From the log, I can see that you are using CUDA 10.0. I'm afraid that it's not yet supported. There are some ambiguities around the templates that use `at:Half`.

**Comment by ygwangmsra** [comment]
Supported.  I have tested the master branch code with v140(VS2015), cuDNN7.3, CUDA10.0, it works.

**Comment by peterjc123** [comment]
@ygwangmsra Alright. Then something about AVX must be changed. Sorry, I don't what happened for newer MSVC versions.

**Comment by ygwangmsra** [comment]
oops. Thanks a lot anyway :-)

**Comment by peterjc123** [comment]
@ygwangmsra What about removing the reference to `C++17.h` in that file first?

**Comment by ygwangmsra** [comment]
not working. the same error information.

**Comment by peterjc123** [comment]
There are two errors that are fatal. I'm not sure which one is the root cause.
```
d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(812): note: See instantiating "at::native::`anonymous" for the class template being compiled -namespace'::ApplyGridSample<scalar_t,2,at::native::detail::GridSamplerInterpolation::Nearest,at::native::detail::GridSamplerPadding::Reflection>"
    14>d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(616): error C2893: The function template "at::vec256::`anonymous- Namespace'::Vec256<int_of_size<sizeof(T)>::type> at::vec256::`anonymous-namespace'::convert_to_int_of_same_size(const at::vec256::`anonymous-namespace'::Vec256<T> &)" specialization [D:\usefullib\pytorch\build\caffe2\caffe2.vcxproj]
         d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(616): note: Use the following template parameters:
         d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(616): note: “T=scalar_t”
    14>d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(620): error C3536: “i_x_nearest”: Cannot be used before initialization [D:\usefullib\pytorch \build\caffe2\caffe2.vcxproj]
    14>d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(620): error C2678: Binary ">": No left operation accepting "int" type Number operator (or no acceptable conversion) [D:\usefullib\pytorch\build\caffe2\caffe2.vcxproj]
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: Probably "at::Tensor at::operator >(at::Scalar,const at::Tensor &)"
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: or "at::Tensor at::operator >(const at::Tensor &,at::Scalar)"
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: or "at::Tensor at::operator >(const at::Tensor &,const at::Tensor &)"
         d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(620): note: Try matching the parameter list "(int, at::vec256::`anonymous-namespace '::Vec256<int32_t>)"
    14>d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(620): error C2678: Binary "<": No left operation accepting "int" type Number operator (or no acceptable conversion) [D:\usefullib\pytorch\build\caffe2\caffe2.vcxproj]
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: Probably "at::Tensor at::operator <(at::Scalar,const at::Tensor &)"
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: or "at::Tensor at::operator <(const at::Tensor &,at::Scalar)"
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: or "at::Tensor at::operator <(const at::Tensor &,const at::Tensor &)"
         d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(620): note: Try matching the parameter list "(int, at::vec256::`anonymous-namespace '::Vec256<int32_t>)"
    14>d:\usefullib\pytorch\build\aten\src\aten\native\cpu\gridsamplerkernel.cpp.avx2.cpp(623): error C2678: Binary "+": not found accept "at::vec256:: The operator of the left operand of the `anonymous-namespace'::Vec256<int32_t>" type (or no acceptable conversion) [D:\usefullib\pytorch\build\caffe2\caffe2.vcxproj]
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: Probably "at::Tensor at::operator +(at::Scalar,const at::Tensor &)"
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: or "at::Tensor at::operator +(const at::Tensor &,at::Scalar)"
         d:\usefullib\pytorch\aten\src\aten\tensoroperators.h(93): note: or “at::Tensor at::operator +(const at::Tensor &,const at::Tensor &)”
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(207): note: or "at::Half at::operator +(long,at::Half)"
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(194): note: or "at::Half at::operator +(at::Half,long)"
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(180): note: or "at::Half at::operator +(int,at::Half)"
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(167): note: or "at::Half at::operator +(at::Half,int)"
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(152): note: or "double at::operator +(double,at::Half)"
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(139): note: or "double at::operator +(at::Half,double)"
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(111): note: or “float at::operator +(float,at::Half)”
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(98): note: or “float at::operator +(at::Half,float)”
         d:\usefullib\pytorch\aten\src\aten\core\half-inl.h(56): note: or "at::Half at::operator +(const at::Half &,const at::Half &)"
```
I don't know what type `scalar_t` is in the first error. While the second message implies that we are missing some templates for the given types <int, at:Vec256<int32_t>>, I don't know if it is the case. If it is the case, why doesn't it throw on earlier MSVC versions?

**Comment by peterjc123** [comment]
@ygwangmsra Would you please try adding the following piece of code near line 616 in `aten\src\aten\native\cpu\gridsamplerkernel.cpp`.
```cpp
    auto x_nearest = x.round();
    auto y_nearest = y.round();

// CODE START

    static constexpr int size = Vec256<scalar_t>::size;
    scalar_t src_arr[size];
    src.store(static_cast<void*>(src_arr));
    int_same_size_t<scalar_t> buffer[size];
    for (int64_t i = 0; i < size; i++) {
      buffer[i] = static_cast<int_same_size_t<scalar_t>>(src_arr[i]);
    }

    auto result = Vec256<int_same_size_t<scalar_t>>::loadu(static_cast<void*>(buffer));

// CODE END

    auto i_x_nearest = convert_to_int_of_same_size(x_nearest);
    auto i_y_nearest = convert_to_int_of_same_size(y_nearest);
```

**Comment by ygwangmsra** [comment]
@peterjc123 I think I have finnally found the problem and solved it. The bug may comes from the VS2017 compiler. Similar case have be reported in [https://developercommunity.visualstudio.com/content/problem/75357/c-compiler-regression-cannot-compile-my-code-with.html](url)

The following is how I solve the problem, may be lucky :-)

Several minutes ago, I have tried the latest VS compiler 1915 (VS2017, update 8) and surprisingly, caffe2 with CUDA can be sucessfully compiled. Although CUDA10 reports that "Microsoft Visual Studio 2017 (RTW and Update 6)" is supported [https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cuda-compiler-new-features](url).

But finnally I found (VS2017 Update 8) is also supported as the host compiler for CUDA10. 

SO, document may be important, misleading document may give us headache. AND missing document is even worse, ummmm, like Caffe2 :-( But, these stuff can not prevent me to use and like this deep learning framework.
